lib/nerves_hub/console_channel.ex

```diff
@@ -0,0 +1,175 @@
+defmodule NervesHub.ConsoleChannel do
+  @moduledoc """
+  Wraps IEx.Server setting this process as the group_leader.
+  This allows tunneling the `io_request`s over the channel socket.
+
+  The basic layout is:
+
+  `NervesHub.ConsoleChannel` (this process) ==> `NervesHubDeviceWww.ConsoleChannel`
+                                                        ||
+                                                        ||
+                                                        ||
+                                                        \\/
+                                              NervesHubCore.DeviceConsole
+                                                        /\\
+                                                        ||
+                                                        ||
+                                                        ||
+  `console.html` (javascript/frontend) =======> `NervesHubWebWww.ConsoleChannel`
+
+  Commands are sent over their respective socket and coordinated via NervesHubCore.DeviceConsole.
+
+  commands are shaped like:
+    * event - "io_request:" <> kind
+      * kind is any of the commands handled by a `group_leader`
+        At least these need to be implemented:
+        * `put_chars`
+        * `get_line`
+    * payload - `%{"data" => data}`
+      * `data` varies per `kind`.
+        * `put_chars` request data is a binary to be written to the console.
+        * `put_chars` response is an `"ok"` or `"error"` binary.
+        * `get_chars` request data is a binary to prompt on the console.
+        * `get_chars` response data is a binary to be processed by `IEx.Server`.
+  """
+  use PhoenixChannelClient
+
+  def topic do
+    "console:" <> Nerves.Runtime.KV.get("nerves_serial_number")
+  end
+
+  def iex_init_callback() do
+    :ok
+  end
+
+  def init_iex(state) do
+    true = :erlang.group_leader(self(), self())
```

Comment:
Setting group leader to itself is a great way to deadlock a process 😅

Basically operations like `IO.puts` or `io:write` send a message to the group leader and await response, but since it's the same process the response will never arrive.

I would suggest spawning a separate process from the channel to actually handle the io as an intermediary between iex and the channel.

Reply:
Thanks for the suggestion! When first implementing this it felt weird setting a group leader to itself. I did notice weird behaviour doing `IO.puts` when debugging this module so that is probably what happened.